testplan/web_ui/testing/src/state/appSlice.js

```diff
@@ -0,0 +1,85 @@
+// @ts-nocheck
+import { createSlice } from '@reduxjs/toolkit/dist/redux-toolkit.esm';
+import { headers as DefaultHeaders } from 'axios/lib/defaults';
+import { flattened } from '../Common/utils';
+
+const __DEV__ = process.env.NODE_ENV !== 'production';
+const API_BASE_URL = process.env.REACT_APP_API_BASE_URL;
+
+/** always false in production */
+const boolIfDev = val => __DEV__ ? !!val : false;
+
+/**
+ * This state slice contains app-wide information that could be used in any
+ * part of the app
+ */
+export default (() => {
+  let apiBaseUrlObj = null;
+  if(typeof API_BASE_URL === 'string' && API_BASE_URL.length) {
+    try {  // try to parse the URL as-is
+      apiBaseUrlObj = new URL(API_BASE_URL);
+    } catch(_) {  // containing "/" means it's a path (fine) or protocol (bad)
+      if(API_BASE_URL.includes('/')) {
+        try {
+          apiBaseUrlObj = new URL(API_BASE_URL, window.location.origin);
+        } catch(_) {
+          // if we get here then the environment variable REACT_APP_API_BASE_URL
+          // has been defined to a non-url value and we cannot go on...
+        }
+      } else {
+        // if we get here then it's impossible to determine whether the
+        // variable's meant to be a hostname or a pathname, and so we again
+        // cannot go on...
+      }
+      if(apiBaseUrlObj === null) {
+        const baseErrMsg = `Cannot compute a URL from "${API_BASE_URL}"`;
+        if(__DEV__) throw new Error(flattened`
+          ${baseErrMsg}, change the environment variable 
+          REACT_APP_API_BASE_URL to the form "/[path]" or 
+          "protocol://hostname[:port]/[path]" for a different host. (note 
+          "[...]" denotes optional components - don't actually include them 
+          in the URL)
+        `);
+        throw new Error(baseErrMsg);
+      }
+    }
+  } else {  // fall back to the root URL path
+    apiBaseUrlObj = new URL('/', window.location.origin);
+    if(__DEV__) console.warn(flattened`
+      Using "${apiBaseUrlObj.href}" as the API URL. Define the environment 
+      variable REACT_APP_API_BASE_URL to use something different.
+    `);
+  }
+
+  const apiHeaders = { ...DefaultHeaders.common, Accept: 'application/json' };
+  if(__DEV__ && window.location.origin !== apiBaseUrlObj.origin) {
+    // CORS headers: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+    apiHeaders['Access-Control-Allow-Origin'] = apiBaseUrlObj.origin;
+  }
+
+  return createSlice({
+    name: 'app',
+    initialState: {
+      apiHeaders,
+      skipFetch: false,
+      apiBaseURL: apiBaseUrlObj.href,
+      isTesting: process.env.NODE_ENV === 'test',
+      isDevel: process.env.NODE_ENV === 'development',
```

Comment:
isTesting/isDevel -- do we need these two states,  considering we already have DEV.

Reply:
it makes sense when I have the URL params connector, which I've removed for this PR. I'll remove this until I add that connector back.